File: examples/ai-pipeline-python/kotlin/src/main/kotlin/dev/pykotmig/aipipeline/Chain.kt
package dev.pykotmig.aipipeline

import dev.langchain4j.data.document.Document

/**
 * Composable pipeline mirroring the Python LangChain `Runnable` demo (normalize → retrieve → render)
 * without live LLM calls. Retrieval uses LangChain4j [Document] metadata; swap [renderAnswer] for a
 * [dev.langchain4j.model.chat.ChatLanguageModel] when wiring production RAG.
 */
fun buildDemoPipeline(): (Map<String, Any?>) -> Map<String, Any?> =
    { input ->
        renderAnswer(buildContext(normalizeQuery(input)))
    }

private fun normalizeQuery(data: Map<String, Any?>): Map<String, Any?> {
    val q = (data["query"] as? String ?: "").trim()
    return mapOf("query" to q)
}

private fun buildContext(data: Map<String, Any?>): Map<String, Any?> {
    val docs =
        listOf(
            Document.from("overview"),
            Document.from("faq"),
        )
    val ctx = docs.map { "doc:${it.text()}" }
    return data + mapOf("context" to ctx)
}

private fun renderAnswer(data: Map<String, Any?>): Map<String, Any?> {
    val q = data["query"] as String
    @Suppress("UNCHECKED_CAST")
    val ctx = data["context"] as? List<*> ?: emptyList<Any>()
    val answer = "q='$q' sources=${ctx.size}"
    return mapOf("answer" to answer)
}

File: examples/ai-pipeline-python/python/src/ai_pipeline_demo/chain.py
"""Composable Runnable pipeline (RAG/agent mental model, no network)."""

from __future__ import annotations

from typing import Any

from langchain_core.runnables import RunnableLambda


def _normalize_query(data: dict[str, Any]) -> dict[str, Any]:
    q = str(data.get("query", "")).strip()
    return {"query": q}


def _build_context(data: dict[str, Any]) -> dict[str, Any]:
    """Fake retriever: attach static context lines."""
    ctx = ["doc:overview", "doc:faq"]
    return {**data, "context": ctx}


def _render_answer(data: dict[str, Any]) -> dict[str, Any]:
    q = data["query"]
    ctx = data.get("context", [])
    return {"answer": f"q={q!r} sources={len(ctx)}"}


def build_demo_pipeline() -> Any:
    """Return a linear chain: normalize → retrieve → render."""
    return (
        RunnableLambda(_normalize_query)
        | RunnableLambda(_build_context)
        | RunnableLambda(_render_answer)
    )
